### DISCUSSION TITLE
What's the right way to insert code after validation?

### DISCUSSION THREAD

**Discussion by jiashengguo** [eclipse-langium/langium#880] [Q&A] [answered]
For example,  for the DomainModel,  after user input 
```

entity Post{
many comments: Comment
}
```
Instead of showing the error of not finding a Comment,  I would like to generate a Comment entity for the user if he hasn't done so.  

How should I achieve that?

**Comment by msujew** [comment, marked as answer]
Hey @jiashengguo,

to not interrupt the [normal document lifecycle](https://langium.org/docs/document-lifecycle/), it would make sense to basically do the following:

1. Register a [`onBuildPhase`](https://github.com/langium/langium/blob/feae27c6fbc3462ac39d60f0e6c570b19821f6f7/packages/langium/src/workspace/document-builder.ts#L58) listener on your document builder instance for the `Validated` state (i.e. so that the callback runs after a document has been validated)
2. Look in the `references` of your document for unresolved references (i.e. those reference with their `error` not being undefined)
3. Create a new workspace edit for all missing elements and send them to the client using the [`workspace/applyEdit`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_applyEdit) LSP endpoint available on the `services.lsp.Connection` object.

Note that **I would not recommend** building such an automated solution due to false positives. Let's take the following for example:

```
entity Post {
  many comments: Commet // <- Note the missing `n` character
}
entity Comment { // <- This is what the user wanted to reference
  ...
}
entity Commet { // <- This will be generated automatically instead
  ...
}
```

Once the user writes an invalid reference, it will automatically create a new entity in the text which might not be what the user wants, especially if it's only due to a typo. You should let the user decide whether they want to create an entity for that instead using [code actions](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_codeAction). You can see an example how to create code actions on linking errors [here](https://github.com/langium/langium/blob/feae27c6fbc3462ac39d60f0e6c570b19821f6f7/packages/langium/src/grammar/lsp/grammar-code-actions.ts#L79-L88).

**Reply by jiashengguo** [reply]
@msujew Thanks for the quick response and reminder! Actually, I just use that simple example to illustrate that I need to insert some code after the validation phase. I totally agree with you that it makes more sense to use code action for that case.
